File: src/joy/fluent-move.clj
(ns joy.fluent-move
  (:gen-class))

(defmacro dbg [x] `(let [x# ~x] (println "dbg:" '~x "=" x#) x#))

;; Super simple implementation
{:from "e7", :to "e8", :castle? false, :promotion \Q}

(defn build-move [& pieces]
  (apply hash-map pieces))

(build-move :from "e7" :to "e8" :promotion \Q)
;=> {:from "e7", :to "e8", :promotion \Q}



;; Implemntation using records
(defrecord Move [from to castle? promotion]
  Object
  (toString [this]
    (str "Move " (:from this)
         " to " (:to this)
         (if (:castle? this) " castle"
             (if-let [p (:promotion this)]
               (str " promote to " p)
               "")))))

(.println System/out (Move. "e7" "e8" nil \Q))



;; Implementation that seperates concerns (i.e. Move record does not need to worry about what is a legal set of params)
(defn build-move [& {:keys [from to castle? promotion]}]
  {:pre [from to]}
  (Move. from to castle? promotion))

(str (build-move :from "e2" :to "e4"))
;=> "Move e2 to e4"

(true? [true true])

File: src/joy/FluentMove.java
package joy;

public class FluentMove {
    String from, to, promotion = "";
    boolean castlep;

    public static MoveBuilder desc() { return new MoveBuilder(); }

    public String toString() {
        return "Move " + from +
            " to " + to +
            (castlep ? " castle" : "") +
            (promotion.length() != 0 ? " promote to " + promotion : "");
    }

    public static final class MoveBuilder {
        FluentMove move = new FluentMove();

        public MoveBuilder from(String from) {
            move.from = from; return this;
        }

        public MoveBuilder to(String to) {
            move.to = to; return this;
        }

        public MoveBuilder castle() {
            move.castlep = true; return this;
        }

        public MoveBuilder promoteTo(String promotion) {
            move.promotion = promotion; return this;
        }

        public FluentMove build() { return move; }
    }

    public static void main(String[] args) {
        FluentMove move = FluentMove.desc()
            .from("e2")
            .to("e4").build();
        System.out.println(move);

        move = FluentMove.desc()
            .from("a1")
            .to("c1")
            .castle().build();
        System.out.println(move);

        move = FluentMove.desc()
            .from("a7")
            .to("a8")
            .promoteTo("Q").build();
        System.out.println(move);
    }
}

